models/Dividends.go

```diff
@@ -0,0 +1,191 @@
+package models
+
+import (
+	"fmt"
+	"github.com/Sirupsen/logrus"
+	"github.com/delta/dalal-street-server/utils"
+)
+
+//UserDetails stores users who own stocks of a given stockid
+type UserDetails struct {
+	UserID        uint32
+	StockQuantity uint64
+}
+
+//PerformDividendsTransaction finds the users who own the stocks of this Company,
+//                            performs dividendTransaction for each of these users,
+//                            updates cash of each of these Users,
+//                            sends the DividendTransactions through datastreams and a notification.
+func PerformDividendsTransaction(stockID uint32, dividendAmount uint64) (string, error) {
+	var l = logger.WithFields(logrus.Fields{
+		"method":               "PerformDividendTransaction",
+		"param_stockId":        stockID,
+		"param_dividendAmount": dividendAmount,
+	})
+
+	var transactions []*Transaction
+
+	var oldCashMap []*User
+
+	l.Infof("PerformDividendTransaction requested")
+
+	db := getDB()
+
+	maxStockId, err := GetMaxStockId()
+
+	l.Errorf("Max stock id : %v", maxStockId)
+
+	if err != nil {
+		l.Errorf("Failure to get max stock id due to : %v+", err)
+		return "Failure", err
+	}
+
+	if stockID < 0 || stockID > maxStockId {
+		l.Errorf("Failure with stock id : %v", stockID)
+		return "Failure", InvalidStockIdError{}
+	}
+
+	/* Committing to database */
+	tx := db.Begin()
+
+	errorHelper := func(format string, args ...interface{}) (string, error) {
+		l.Errorf(format, args...)
+		tx.Rollback()
+		return "failure", fmt.Errorf(format, args...)
+	}
+
+	var benefittingUsers []*UserDetails
+
+	db.Raw("SELECT userId As user_id ,SUM(stockQuantity) AS stock_quantity FROM Transactions WHERE stockId = ? GROUP BY userId HAVING SUM(stockQuantity) > 0 ", stockID).Scan(&benefittingUsers)
+
+	l.Infof("Successfully fetched users who own stocks of stockID : %v", stockID)
+
+	for _, user := range benefittingUsers {
+
+		dividendTotal := dividendAmount * uint64(user.StockQuantity)
+		l.Infof(" The user id is %v and stock quantity is : %v", user.UserID, user.StockQuantity)
+
+		transaction := &Transaction{
+			UserId:        user.UserID,
+			StockId:       stockID,
+			Type:          DividendTransaction,
+			StockQuantity: int64(user.StockQuantity),
```

Comment:
Should be zero as the stocks for which user is paid dividends is doubled every time